### PR TITLE
Feature/gamejam

### PR DESCRIPTION
**Pull request by GavanWilhite**

### CODE REVIEW

File: examples/models/chess/.gitignore [hunk 1]
```diff
@@ -0,0 +1 @@
+*.obj.cs
```
Comment (brianpeiris) [inline examples/models/chess/.gitignore]: What's this for?

File: src/utilities/behaviors/Elemental.js [hunk 2]
```diff
@@ -0,0 +1,134 @@
+window.altspace = window.altspace || {};
+window.altspace.utilities = window.altspace.utilities || {};
+window.altspace.utilities.behaviors = window.altspace.utilities.behaviors || {};
+
+/**
+ * A behavior that renders out a three.js scene to DOM elements. This allows for easier debugging via the Chrome inspector.
```
Comment (brianpeiris) [inline src/utilities/behaviors/Elemental.js]: This should have a better description of what it does, if we actually expect people to use it. Specifically, mention that it adds an `x-elemental` element to `head` (which I found unexpected). Also probably mention that the behavior should be removed in production.

File: src/utilities/behaviors/Elemental.js [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+        positionEl.setAttribute('x', formatNumber(this.position.x));
+        positionEl.setAttribute('y', formatNumber(this.position.y));
+        positionEl.setAttribute('z', formatNumber(this.position.z));
+
+        var existingRotationEls = el.getElementsByTagName('x-rotation');
+        var rotationEl = existingRotationEls[0] || el.appendChild(document.createElement('x-rotation'));
+        rotationEl.setAttribute('x', formatNumber(this.rotation.x));
+        rotationEl.setAttribute('y', formatNumber(this.rotation.y));
+        rotationEl.setAttribute('z', formatNumber(this.rotation.z));
+        */
```
Comment (brianpeiris) [inline src/utilities/behaviors/Elemental.js]: Kill dead code.

File: src/utilities/behaviors/Elemental.js [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+        return el;
+    }
+
+    function resetRoot() {
+        if (rootEl) rootEl.remove();
+        rootEl = null;
+        rootEl = document.createElement('three-elemental');
+    }
```
Comment (brianpeiris) [inline src/utilities/behaviors/Elemental.js]: This function isn't used.

File: src/utilities/behaviors/Elemental.js [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+window.altspace.utilities = window.altspace.utilities || {};
+window.altspace.utilities.behaviors = window.altspace.utilities.behaviors || {};
+
+/**
+ * A behavior that renders out a three.js scene to DOM elements. This allows for easier debugging via the Chrome inspector.
+ * @class Elemental
+ * @param {Object} [config]
+ */
+altspace.utilities.behaviors.Elemental = function (config) {
+    config = config || {};
```
Comment (brianpeiris) [inline src/utilities/behaviors/Elemental.js]: Use tab indentation.

File: src/utilities/behaviors/Object3DSync.js [hunk 6]
```diff
@@ -66,29 +72,25 @@ window.altspace.utilities.behaviors.Object3DSync = function (config){
             if (config.scale) {
                 object3d.scale.set(value.scale.x, value.scale.y, value.scale.z);
             }
-            if (config.syncData) {
```
Comment (brianpeiris) [inline src/utilities/behaviors/Object3DSync.js]: Docs need to be updated if we've removed this syncData flag.

File: src/utilities/behaviors/Object3DSync.js [hunk 7]
```diff
@@ -66,29 +72,25 @@ window.altspace.utilities.behaviors.Object3DSync = function (config){
...
-     * 
-     * This is to be used whenever you update a property and are not using auto. If multiple users could potentially move an object, this is preferred vs using auto.
-     * @instance
-     * @method enqueueSend
-     * @memberof module:altspace/utilities/behaviors.Object3DSync
-     */
-    function enqueueSend() {
-        sendEnqueued = true;
+            var gained = newOwnerId === sceneSync.clientId && !isMine;
+            if (gained) object3d.dispatchEvent({ type: 'ownershipgained' });
```
Comment (brianpeiris) [inline src/utilities/behaviors/Object3DSync.js]: These new events should be documented.

File: src/utilities/behaviors/Object3DSync.js [hunk 8]
```diff
@@ -144,8 +137,25 @@ window.altspace.utilities.behaviors.Object3DSync = function (config){
...
+        get: function () {
+            return dataRef;
+        }
+    });
+
+    Object.defineProperty(exports, 'isMine', {
+        get: function () {
+            return isMine;
+        }
+    });
```
Comment (brianpeiris) [inline src/utilities/behaviors/Object3DSync.js]: These new properties should be documented.

File: src/utilities/behaviors/Object3DSync.js [hunk 9]
```diff
@@ -144,8 +137,25 @@ window.altspace.utilities.behaviors.Object3DSync = function (config){
...
 
     return exports;
 };
 
+//manual modifications to the ref's will not obey ownership status.
```
Comment (brianpeiris) [inline src/utilities/behaviors/Object3DSync.js]: Put this in the docs?

File: src/utilities/behaviors/SceneSync.js [hunk 10]
```diff
@@ -24,6 +24,8 @@ window.altspace.utilities.behaviors = window.altspace.utilities.behaviors || {};
  **/
 window.altspace.utilities.behaviors.SceneSync = function (instanceRef, config) {
     var sceneRef = instanceRef.child('scene');
+    var clientsRef = instanceRef.child('clients');
```
Comment (brianpeiris) [inline src/utilities/behaviors/SceneSync.js]: Use tabs for indentation.

File: src/utilities/behaviors/SceneSync.js [hunk 11]
```diff
@@ -31,19 +33,76 @@ window.altspace.utilities.behaviors.SceneSync = function (instanceRef, config) {
...
+        clientsRef.on("value", function (snapshot) {
+            var clientIds = snapshot.val();
+
+            if (!clientIds) return;
+
+            masterClientId = clientIds[0];
+        });
+        // add our client ID to the list of connected clients, 
+        // but have it be automatically removed by firebase if we disconnect for any reason
+        clientsRef.push(clientId).onDisconnect().remove();
```
Comment (brianpeiris) [inline src/utilities/behaviors/SceneSync.js]: Awesome, this is super useful.

File: src/utilities/behaviors/SceneSync.js [hunk 12]
```diff
@@ -188,5 +212,11 @@ window.altspace.utilities.behaviors.SceneSync = function (instanceRef, config) {
     Object.defineProperty(exports, 'autoSendRateMS', {
         get: function () { return autoSendRateMS; }
     });
+    Object.defineProperty(exports, 'isMasterClient', {
+        get: function () { return masterClientId === clientId; }
+    });
+    Object.defineProperty(exports, 'clientId', {
+        get: function () { return clientId; }
+    });
```
Comment (brianpeiris) [inline src/utilities/behaviors/SceneSync.js]: Needs documentation.

File: src/utilities/shims/behaviors.js [hunk 13]
```diff
@@ -205,9 +238,15 @@ THREE.Object3D.prototype.updateBehaviors = function(deltaTime, scene) {
...
+        } catch (error) {
 
-            (console.error || console.log).call(console, e.stack || e);
+            console.group();
+            (console.error || console.log).call(console, error.stack || error);
+            console.log('[Behavior]');
+            console.log(behavior);
+            console.log('[Object3D]');
+            console.log(this);
+            console.groupEnd();
```
Comment (brianpeiris) [inline src/utilities/shims/behaviors.js]: Put this in a helper function, since it's repeated several times.

Comment (GavanWilhite) [inline src/utilities/shims/behaviors.js]: Good stuff!

Recall trying to use a helper function and it not being very elegant for
some reason, will have to look.

obj.cs is some stupid thing that gets compiled by VS or prepros (it tries
to do something with.obj files).

Elemental I was a pretty rough test, I'll probably pull it out.

Comment (brianpeiris) [inline src/utilities/shims/behaviors.js]: Sounds good. Probably just add obj.cs to the main .gitignore file.

Comment (GavanWilhite) [inline src/utilities/shims/behaviors.js]: Ah yeah, I think my GUI went a little crazy there

**Pull request merged**